src/memory/windows.rs

```diff
@@ -10,125 +10,136 @@ use winapi::um::winnt::{
+        let mut candidates = Vec::new();
+        let mut base_address = 0;
+        let mut memory_info: MEMORY_BASIC_INFORMATION = unsafe { zeroed() };
 
-    for &address in candidates {
-        let mut bytes_read = 0;
-        let mut value_at_address = T::new();
-        let read_success = unsafe {
-            ReadProcessMemory(
+        while unsafe {
+            VirtualQueryEx(
                 process_handle,
-                address as *const _,
-                &mut value_at_address as *mut _ as *mut _,
-                std::mem::size_of::<i32>(),
-                &mut bytes_read,
+                base_address as *const _,
+                &mut memory_info,
+                size_of::<MEMORY_BASIC_INFORMATION>(),
             )
-        };
-        if read_success != 0 && bytes_read > 0 && value_at_address == value {
-            println!("FOUND {} at {:#x}", value, address);
-            remaining_candidates.push(address);
+        } == size_of::<MEMORY_BASIC_INFORMATION>()
+        {
+            // Only consider regions that are committed, ignoring other states for simplicity
+            if memory_info.State == winapi::um::winnt::MEM_COMMIT {
+                let mut buffer: Vec<u8> = vec![0; memory_info.RegionSize];
+                let mut bytes_read = 0;
+
+                let read_success = unsafe {
+                    ReadProcessMemory(
+                        process_handle,
+                        memory_info.BaseAddress,
+                        buffer.as_mut_ptr() as *mut _,
+                        memory_info.RegionSize,
+                        &mut bytes_read,
+                    )
+                };
+
+                if read_success != 0 && bytes_read > 0 {
+                    candidates.append(&mut first_search(&buffer, value, base_address));
+                }
+            }
+
+            // Move to the next memory region
+            base_address = (memory_info.BaseAddress as usize) + memory_info.RegionSize;
+        }
+        unsafe {
+            CloseHandle(process_handle);
         }
+        candidates
     }
-    remaining_candidates
-}
+
+    fn search_among_candidates(&self, pid: i32, value: T, candidates: &[usize]) -> Vec<usize> {
+        let process_handle =
+            unsafe { OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, 0, pid as u32) };
+
+        if process_handle.is_null() {
+            println!("Failed to open process. Error: {}", unsafe {
+                GetLastError()
+            });
+            return Vec::new();
+        }
+
+        let mut remaining_candidates = Vec::new();
+
+        for &address in candidates {
+            let mut bytes_read = 0;
+            let mut value_at_address = T::new();
+            let read_success = unsafe {
+                ReadProcessMemory(
+                    process_handle,
+                    address as *const _,
+                    &mut value_at_address as *mut _ as *mut _,
+                    size_of::<i32>(),

```

Comment:
This is not related to your PR, but I think this is a bug. It should be `size_of::<T>`. You can fix it, or I can fix it later.

Reply:
Fixed.